keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,660 @@
+extension in the schema. In the above example, the validator is scoped to the
+'spec' field.
+
+- For OpenAPIv3 object types, the expression will have direct access to all the
+fields of the object the validator is scoped to.
+  
+- For OpenAPIv3 scalar types (integer, string & boolean), the expression will have
+access to the scalar data element the validator is scoped to.
+
+- For OpenAPIv3 list and map types, the expression will have access to the data
+element of the list or map.
+  
+- For immutability use case, validator will have access to the existing version of the object.
+
+- We plan to allow access to the current state of the object to allow validation rules to check the new value against the current value(e.g. for validation ratcheting or immutability checks).
+
+
+#### Field paths and field patterns
+
+A field path is a patch to a single node in the data tree. I.e. it specifies the
+exact indices of the list items and the keys of map entries it traverses.
+
+A field *pattern* is a path to all nodes in the data tree that match the pattern. I.e.
+it may wildcard list item and map keys.
+
+
+#### Expression lifecycle
+
+When CRDs are written to the kube-apiserver, all expressions will be [parsed and
+typechecked](https://github.com/google/cel-go#parse-and-check) and the resulting
+program will be cached for later evaluation (CEL evaluation is thread-safe and
+side-effect free). Any parsing or type checking errors will cause the CRD write
+to fail with a descriptive error.
+
+#### Function library
+
+The function library available to expressions can be augmented using [extension
+functions](https://github.com/google/cel-spec/blob/master/doc/langdef.md#extension-functions).
+
+List of functions to include for the initial release:
+- Equality and Ordering (customized to do Kuberenetes semantic equality that handles "associative lists")
+- Regular Expressions
+- Some Standard Definitions
+
+
+Considerations:
+- The functions will become VERY difficult to change as this feature matures. We
+  should limit ourselves initially to functions that we have a high level of
+  confidence will not need to be changed or rethought.
+
+- Support kubernetes specific concepts, like accessing associative lists by key may be needed, but
+  we need to review more use cases to determine if this is needed.
+  
+- The Kubernetes associated list equality uses map semantic equality which is different from CEL. 
+  We would consider overwriting in CEL or adding a workaround utility function.
+
+
+### User Stories
+
+- Use case: [Tekton pipeline validation](https://github.com/tektoncd/pipeline/blob/main/pkg/apis/pipeline/v1beta1/pipeline_validation.go)
+  - Referential integrity checks
+  - Custom formatted validation error messages
+  - "Either list X or list Y must be non-empty"
+  - "There exists a X in one list/map for each Y in another map/list"
+- (PRs to add additional user stories to this list welcome!)
```

Comment:
1. list of type foo struct {name string ... }, no item in the list can have a name == "value X", [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/apiserver/validate_apiserver.go#L68).
2. metadata.name must equal "valueX", [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/apiserver/validate_apiserver.go#L47)
3. if name == "foo", then fieldX must not be nil, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/apiserver/validate_apiserver.go#L177)
4. if name == "foo", then field X must be nil, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/apiserver/validate_apiserver.go#L177)
5. validate new label selector format: metav1.LabelSelector, this matches our deployment API
6. validate old label selector format: map[string]string, this matches our node selectors for pods
7. len(A)>0 xor len(B)>0
8. quantity validation to match quota and usage specifications, [buried inside of here (most of it is not generally applicable)](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/clusterresourcequota/validation/validation.go#L44)
9. if old value is X, don't allow changing the value, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/features/validate_features.go#L82)
10. shorthand to match our NameIsDNSSubdomain and the like would be nice.
11. if X exists in list1, then X must not existing in list2, both list1 and list2 are in the manifest, [ref](https://github.com/openshift/kubernetes/blob/75ee3073266f07baaba5db004cde0636425737cf/openshift-kube-apiserver/admission/customresourcevalidation/securitycontextconstraints/validation/validation.go#L139)
12.

Reply:
Huge thanks. Added